dns/resolver.go

```diff
@@ -0,0 +1,385 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dns
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"math/rand"
+	"net"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+	"github.com/Jigsaw-Code/outline-sdk/transport/tls"
+	"golang.org/x/net/dns/dnsmessage"
+)
+
+var (
+	ErrBadRequest  = errors.New("request input is invalid")
+	ErrDial        = errors.New("dial DNS resolver failed")
+	ErrSend        = errors.New("send DNS message failed")
+	ErrReceive     = errors.New("receive DNS message failed")
+	ErrBadResponse = errors.New("response message is invalid")
+)
+
+type nestedError struct {
```

Comment:
I added a comment.

I'd rather not expose a new type we need to maintain for now, since it's not needed. We may want to move it to internal and expose to the SDK if we want to use in other places in the future.